tensorflow/core/kernels/scatter_functor.h

```diff
@@ -205,17 +207,61 @@ struct ScatterFunctorBase {
     // indices and params sizes were validated in DoCompute().
     const Index N = static_cast<Index>(indices.size());
     const Index limit = static_cast<Index>(params.dimension(0));
-    for (Index i = 0; i < N; i++) {
-      // Grab the index and check its validity.  Do this carefully,
-      // to avoid checking the value and grabbing it again from
-      // memory a second time (a security risk since it may change in between).
-      const Index index = ::tensorflow::internal::SubtleMustCopy(indices(i));
-      if (!FastBoundsCheck(index, limit)) return i;
-      // Copy last Ndim-1 dimensions of updates[i] to params[index]
-      scatter_op::internal::Assign<op>::Run(params.template chip<0>(index),
-                                            updates.template chip<0>(i));
+    unsigned long int num_locks, entries_per_lock;
+    // Duplicate entries need to be handled correctly.
+    // Multiple updates to the same index has to be serialized.
+    // To reduce the number of locks and the memory usage,
+    // we divide the whole index space into kMaxLocks regions
+    // with each lock serializing access to a region.
+    if (limit <= kMaxLocks) {
+      num_locks = limit;
+      entries_per_lock = 1;
+
+    } else {
+      num_locks = kMaxLocks;
+      entries_per_lock = (limit % kMaxLocks == 0) ? limit / kMaxLocks
+                                                  : (limit / kMaxLocks + 1);
     }
-    return -1;
+
+    std::vector<std::atomic<bool>> accessed(num_locks);
```

Comment:
How much does the performance differ from `std::atomic_flag`? Could you please try [std::atomic_flag::test_and_set](http://www.cplusplus.com/reference/atomic/atomic_flag/test_and_set/) as well?

Reply:
Would love to know how TF's `mutex` performs too.

Reply:
Do you mean fine grained locking with tf's mutex?

Reply:
Yes, just curious.

Reply:
Mutex
BM_ScatterAddInt32/1      12911020        100    77.5M items/s
BM_ScatterAddInt32/10     14440060        100    692.5M items/s
BM_ScatterAddInt32/64     19285140        100    3318.6M items/s
BM_ScatterAddInt32/256    43886400        100    5833.2M items/s
BM_ScatterAddInt32/1024  129769780        100    7890.9M items/s

atomic_flag.test_and_set()
BM_ScatterAddInt32/1      19393250        100    51.6M items/s
BM_ScatterAddInt32/10     23358120        100    428.1M items/s
BM_ScatterAddInt32/64     32073190        100    1995.4M items/s
BM_ScatterAddInt32/256    46956580        100    5451.8M items/s
BM_ScatterAddInt32/1024  130544220        100    7844.1M items/s

Current
BM_ScatterAddInt32/1      27138230        100    36.8M items/s
BM_ScatterAddInt32/10     30097940        100    332.2M items/s
BM_ScatterAddInt32/64     35067340        100    1825.1M items/s
BM_ScatterAddInt32/256    54302060        100    4714.4M items/s
BM_ScatterAddInt32/1024  129844650        100    7886.3M items/s

Fine grained mutex seems to perform best overall. However in a model we have the difference between the 3 is not as much pronounced.

Reply:
I can change the code to either. If we are going to opt for mutex option, I would like to know how the GUARDED_BY/EXCLUSIVE_LOCKS syntax works since it is fine grained locking. Please let me know.

Reply:
This is great! Thank you very much for trying!
Since `Mutex` doesn't hurt your model either, let's use `Mutex`. 

I don't think any of the [thread-safety annotations](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/default/thread_annotations.h) are applicable here. `GUARDED_BY` is used when the variable is declared.  We can't really guard `params` as it is passed into the functor. `EXCLUSIVE_LOCKS_REQUIRED` and `SHARED_LOCKS_REQUIRED` means locks must have been acquired before the function, which is not suitable for our case. 

`ACQUIRE_SHARED` and `RELEASE` can't be used together so I don't think we can tag that on the lambda function either.

Reply:
Thanks for the clarification! Do you have an insight of where the perf difference is coming from?

Reply:
I'm guessing that the results are with INTRA_OP=13? Do you see much difference with INTRA_OP=1 or 2? If not, then it's probably because of a more careful implementation to avoid contention between threads.

TensorFlow's [Mutex](https://github.com/tensorflow/tensorflow/blob/9adba1c5ea0eee85dd215a9e9be51a009ce771f2/tensorflow/core/platform/default/mutex.cc#L43) uses [nsync's mutex](https://github.com/google/nsync/blob/master/internal/mu.c). Threads can sleep and wake up every once in a while to check on a lock to avoid contention.

A downside of `while(test)` following by`while(test-and-set)` is that the threads can still contend at `while(test-and-set)` (many threads getting through `while(test)` at the same time).

Reply:
Thanks for the explanation!